### PR TITLE
add X-Sentry-Environment header to email notifications

### PR DESCRIPTION
**Pull request by Code0x58**
What is says on the tin

### CODE REVIEW

File: src/sentry/plugins/sentry_mail/models.py [hunk 1]
```diff
@@ -23,6 +23,7 @@
 from sentry.plugins import register
 from sentry.plugins.base.structs import Notification
 from sentry.plugins.bases.notify import NotificationPlugin
+from sentry.tasks.unmerge import get_environment_name
```
Comment (mattrobenolt) [inline src/sentry/plugins/sentry_mail/models.py]: @tkaemming Is there a better way to derive this? I don't like pulling this random function out of `sentry.tasks.unmerge`.

Comment (tkaemming) [reply src/sentry/plugins/sentry_mail/models.py]: There's not really a clean way to do it — the options I can think of are either importing it here or duplicating the logic (it's not complex.)

I'd considered adding it as a property on the `Event` model but since that requires the `data` field to be populated, the possibility of enabling someone to accidentally implicitly fire off N serial queries in a loop made me nervous, so I didn't.

**Comment by tkaemming** [comment]
I think this change is generally fine, but I'd like to verify that the environment name is limited to valid characters ("A field body may be composed of any US-ASCII characters, except for CR and LF." - [RFC 2822 2.2](https://tools.ietf.org/html/rfc2822#section-2.2)) since validation around this field can be a little hard to follow.

**Comment by Code0x58** [comment]
[RFC5335](https://tools.ietf.org/html/rfc5335#section-4) from 2008 extends SMTP header bodies to contain UTF-8, it looks like at least the Subject header can already contain UTF-8 as it comes from [`Group.get_email_subject(…)`](https://github.com/getsentry/sentry/blob/8.22.0/src/sentry/models/group.py#L399) - I'd guess the other headers could contain unicode too but I haven't looked.

As for CR-LF, I don't know how they are handled in `MessageBuilder` (which appears to extends on Django's code), I could have a look later.

**Comment by tkaemming** [comment]
Thanks for the correction about RFC 5335 — my mistake.

Right now, it looks like the `environment` attribute accepts values with any valid string of unicode characters, as long as the string isn't too long (and I've found several rows with `\n` in our own database): https://github.com/getsentry/sentry/blob/c3c3403088c8dd20200ed273119d945b8ae0c3e0/src/sentry/coreapi.py#L774-L784

I think we'll want to fix the validation before this is ready to be merged, but otherwise I think this patch shouldn't require any changes independently. Thanks!

**Comment by Code0x58** [comment]
After having it look, I think [it would raise a `BadHeaderError`](https://github.com/django/django/blob/1.6.11/django/core/mail/message.py#L81) if the header contained CR or LF; I guess you'd be aware of it by now if any of the other headers could end up containing either of those.

**Comment by Code0x58** [comment]
Since #9815 no more CR or LF characters will appear in new environments, so I'm thinking this can now move ahead.

Would this need to be released some time after the updated environment validation, to avoid trying to email about an event with an old bad environment?

**Review by tkaemming — APPROVED** [review]
This seems good to me. I don't think we need to wait, as all new events that would trigger an alert will have validated environment names. Thanks for the reminder, and for your patience.

**Comment by tkaemming** [comment]
Unfortunately, we had to revert this (see GH-10772.) There are cases where environment values are not being properly validated: https://sentry.io/share/issue/598204f35dfe4daeb78b671417e932cf/
